cpp/velox/substrait/VeloxSubstraitSignature.cc

```diff
@@ -159,16 +151,43 @@ TypePtr VeloxSubstraitSignature::fromSubstraitSignature(const std::string& signa
         }
       }
       types.emplace_back(fromSubstraitSignature(typeStr));
-      names.emplace_back("");
       childrenTypes.erase(0, endPos + delimiter.length());
     }
     if (childrenTypes.size() > 0 && !startWith(childrenTypes, ">")) {
       types.emplace_back(fromSubstraitSignature(childrenTypes));
-      names.emplace_back("");
+    }
+    return types;
+  };
+
+  if (startWith(signature, "dec")) {
+    // Decimal type name is in the format of dec<precision,scale>.
+    auto precisionStart = signature.find_first_of('<');
+    auto tokenIndex = signature.find_first_of(',');
+    auto scaleEnd = signature.find_first_of('>');
+    auto precision = stoi(signature.substr(precisionStart + 1, (tokenIndex - precisionStart - 1)));
+    auto scale = stoi(signature.substr(tokenIndex + 1, (scaleEnd - tokenIndex - 1)));
+    return DECIMAL(precision, scale);
+  }
+
+  if (startWith(signature, "struct")) {
+    // Struct type name is in the format of struct<T1,T2,...,Tn>.
+    auto types = parseNestedTypeSignature(signature);
+    std::vector<std::string> names(types.size());
+    for (int i = 0; i < types.size(); i++) {
+      names[i] = "";
     }
     return std::make_shared<RowType>(std::move(names), std::move(types));
   }
 
+  if (startWith(signature, "map")) {
+    // Map type name is in the format of map<T1,T2>.
+    auto types = parseNestedTypeSignature(signature);
+    if (types.size() != 2) {
+      VELOX_UNSUPPORTED("Substrait type signature conversion to Velox type not supported for {}.", signature);
+    }
+    return MAP(std::move(types)[0], std::move(types)[1]);
+  }
+
   if (startWith(signature, "list")) {
     auto listStart = signature.find_first_of('<');
```

Comment:
@WangGuangxin, it seems we can also use `parseNestedTypeSignature` to handle list type. Right?

Reply:
Seems that the `parseNestedTypeSignature` can't properly handle list type, since `list` type's child doesn't have comma, so for types like `list<map<str,str>>`  it will not work

Reply:
@WangGuangxin, I see. Thanks!